Data/HashMap/Internal/Strict.hs

```diff
@@ -306,9 +306,18 @@ update f = alter (>>= f)
 -- @
 alter :: (Eq k, Hashable k) => (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
 alter f k m =
-  case f (HM.lookup k m) of
-    Nothing -> HM.delete k m
-    Just v  -> insert k v m
+    let !h = hash k
+        !lookupRes = HM.lookupRecordCollision h k m
+    in case f (HM.lookupResToMaybe lookupRes) of
+        Nothing -> case lookupRes of
+            Absent            -> m
+            Present _ collPos -> HM.deleteKeyExists collPos h k m
+        Just !v' -> case lookupRes of
+            Absent             -> HM.insertNewKey h k v' m
+            Present !v collPos ->
+                if v `ptrEq` v'
+                    then m
+                    else HM.insertKeyExists collPos h k v' m
```

Comment:
I wonder whether it is necessary to force the `v` contained in the `Present` result here. If it is the same as `v'`, shouldn't forcing `v'` be sufficient?!

Note that this is the strict API, so `v` should already be in WHNF.

Reply:
People may want to use this strict `alterF` on a `HashMap` that contains value thunks though. I think it shouldn't hurt much to keep the bang on `v`.